doc/stdenv/platform-notes.chapter.md

```diff
@@ -46,22 +46,189 @@ Some common issues when packaging software for Darwin:
+  | 12.0–12.5.1        | 11.3                                              | `apple-sdk_11`    |
+  | 13.0–13.4.1        | 12.3                                              | `apple-sdk_12`    |
+  | 14.0–14.3.1        | 13.3                                              | `apple-sdk_13`    |
+  | 15.0–15.4          | 14.4                                              | `apple-sdk_14`    |
+  | 16.0               | 15.0                                              | `apple-sdk_15`    |
+
+  To use a non-default SDK, add it to your build inputs.
+
+  ```nix
+  stdenv.mkDerivation {
+    name = "libfoo-1.2.3";
+    # ...
+    buildInputs = [ apple-sdk_15 ]; # Uses the 15.0 SDK instead of the default SDK for the platform.
+  }
+  ```
+
+  If your derivation has multiple SDKs its inputs (e.g., because it has been propagated by one your dependencies), it will use the highest SDK version available.
+
+  ```nix
+  stdenv.mkDerivation {
+    name = "libfoo-1.2.3"; # Upstream specifies that it needs Xcode 12 to build, so use the 11.3 SDK.
+    # ...
+    buildInputs = [ apple-sdk_11 ];
+    nativeBuildInputs = [ swift ]; # Propagates the 13.3 SDK, so the 13.3 SDK package will be used instead of the 11.3 SDK.
+  }
+  ```
+
+- You should rely on the default SDK when possible. If a package specifies a required SDK version, use that version (e.g., libuv requires 11.0, so it should use `apple-sdk_11`). When a package supports multiple SDKs, determine which SDK package to use based on the following rules of thumb:
+
+  - If a package supports multiple SDK versions, use the lowest supported SDK version by the package (but no lower than the default SDK). That ensures maximal platform compatibility for the package.
+
+  - If a package specifies a range of supported SDK versions _and_ a minimum supported version, assume the package is using availability checks to support the indicated minimum version.Add latest supported SDK package and a `darwinMinVersionHook` set to the minimum supported by the upstream package.
+
+  Warning: Avoid using newer SDKs than an upstream package supports. When a binary is linked on Darwin, the SDK version used to build it is recorded in the binary. Runtime behavior can vary based on the SDK version, which may work fine but can also result in unexpected behavior or crashes when building with an unsupported SDK.
+
+  ```nix
+  stdenv.mkDerivation {
+    name = "foo-1.2.3";
+    # ...
+    buildInputs = [ apple-sdk_15 (darwinMinVersionHook "10.15") ]; # Upstream builds with the 15.0 SDK but supports 10.15.
+  }
+  ```
+
+- Libraries that require a minimum version can propagate an appropriate SDK and `darwinMinVersionHook`. Derivations using that library will automatically use an appropriate SDK and minimum version. Even if the library builds with a newer SDK, it should propagate the minimum supported SDK. Derivations that need a newer SDK can add it to their `buildInputs`.
```

Comment:
This is a suggestion that libraries can propagate their required SDK/version to derivations that depend on them.

For example: https://github.com/NixOS/nixpkgs/blob/723e1a67de2e381de6b5366e6162f237b1dd7be5/pkgs/development/libraries/qt-6/default.nix#L34-L41

Per the linked documentation, Qt 6 supports macOS 12–15 and supports building with the 15.0 SDK. Qt itself is built with the 15.0 SDK, but it propagates a `darwinMinVersionHook` of 12.0 for the minimum version and the corresponding `apple-sdk_12`. If you use Qt, you get the minimum supported SDK and minimum verison that you need.

The primary value of setting a deployment target is it eliminates the requirement to check for any API of that version or older. If you want to use an API from 11.3 when using a 12.0 deployment target, you just can. Otherwise, if you want to use a newer API, you should check whether it’s available. (It’s an error not to do it in Swift but only a warning in C and C++.)

https://developer.apple.com/tutorials/app-dev-training/adopting-new-api-features